src/aws/nodemanager/nodemanager.py

```diff
@@ -113,15 +136,35 @@ def generate_heartbeat(self, notify=True):
     def process_heartbeat(self, hb, source) -> Packet:
         hb['source'] = source  # Set the source IP of the heartbeat (i.e., the worker).
         self.notify(hb)  # Forward the heartbeat to the monitor for metrics.
-        if len(self._tasks)>0:  #If there are tasks in the taskpool send a new command to the worker
+
+
+        if hb['instance_id'] in list(self._task_assignment.keys()) and not hb['instance_id'] in list(self._tasks_running.keys()):
             packet=CommandPacket(command="task")
-            current_task=self._tasks.pop(0)
-            packet["task_data"]= current_task.get_task_data()
-            current_task.state=TaskState.RUNNING
-            self._tasks_running.append(current_task)
+            packet['task']=self._task_assignment[hb['instance_id']].pop(0)
+            self._task_processing[hb['instance_id']]=[packet['task']]
             return packet
+       
+        # When a new instance is available
+        elif not hb['instance_id'] in self._task_assignment.keys():
+            packet=CommandPacket(command="task")
+            # No tasks are available so steal a pair of tasks
+            if len(self._tasks)==0:
+                stolen_task=steal_task()
+                self._task_processing[hb['instance_id']]=[stolen_task]
+                packet["task"]= stolen_task
+                self._task_assignment[hb['instance_id']]=[steal_task()]
```

Comment:
This is the second steal, is this correct? I think this should be `stolen_task` as well

Reply:
Yes this is a second steal. This wa the worker has 1 task to process and 1 task which it is assigned to after it is done.

Reply:
Ah okay gotcha